MapboxNavigationUI/RouteMapViewController.swift

```diff
@@ -240,7 +240,46 @@ extension RouteMapViewController: NavigationMapViewDelegate {
             }
         }
         
-        return CLLocation(coordinate: newCoordinate, altitude: location.altitude, horizontalAccuracy: location.horizontalAccuracy, verticalAccuracy: location.verticalAccuracy, course: location.course, speed: location.speed, timestamp: location.timestamp)
+        let defaultReturn = CLLocation(coordinate: newCoordinate, altitude: location.altitude, horizontalAccuracy: location.horizontalAccuracy, verticalAccuracy: location.verticalAccuracy, course: location.course, speed: location.speed, timestamp: location.timestamp)
+        
+        guard location.course == -1 else {
+            return defaultReturn
+        }
+        
+        guard let coords = routeController.routeProgress.route.coordinates else {
+            return defaultReturn
+        }
+        
+        guard let closest = closestCoordinate(on: coords, to: location.coordinate) else {
```

Comment:
For a long route, a linear search through the entire route geometry is bound to be slow and pointless, and doing so once a second is could hasten the battery’s demise. On other hand, if we only look at the current step’s geometry, we may miss a closer coordinate along the previous or next step.

So figure out the closest coordinate on the current step, the previous step, and the next step, then take the one nearest to the user’s location. This way we only consider relevant steps, and we also avoid snapping to the wrong portion of the route when the route doubles back on itself, as seen in mapbox/navigation.js#12.

(×3)